### PR TITLE
feat: rename files TDE-497

### PR DESCRIPTION
**Pull request by paulfouquet**
## Description
This PR implements the ability of renaming a file in both local and `s3` file system. The functionality is renaming a path to another, not only the file name. For `s3` the files have to be on the same bucket.

_**EDIT:** We don't need that at the moment. Keep this as a Draft for a bit._

### CODE REVIEW

File: scripts/files/fs_local.py [hunk 1]
```diff
@@ -20,3 +25,30 @@ def read(path: str) -> bytes:
...
+
+
+def rename(path: str, new_path: str) -> None:
+    """Rename a file.
+
+    Args:
+        path (str): original path
+        new_path (str): path to be renamed with
+    """
+    if not os.path.exists(path):
```
Comment (blacha) [inline scripts/files/fs_local.py:37]: checking if a file exists before doing something with it is somewhat of a antipattern.

between the time you check and the time you access it, the file could have been moved.

you are best to handle it in the actual action `os.rename` (Im assuming it throws if the file doesnt exist?

does it also throw if the target does exist?

File: scripts/files/fs_s3.py [hunk 2]
```diff
@@ -67,6 +67,96 @@ def read(path: str, needs_credentials: bool = False) -> bytes:
...
+                destination=new_path,
+                error="A file for the destination path already exists.",
+            )
+            raise Exception(f"{new_path} already exists. The file can't be renamed with this path.")
+        except botocore.exceptions.ClientError as e:
+            if e.response["Error"]["Code"] == "404":
+                # OK: the destination file does not already exists
+                dst_s3_object.copy({"Bucket": src_s3_path.bucket, "Key": src_s3_path.key})
+                # delete the source
+                src_s3_object = s3.Object(src_s3_path.bucket, src_s3_path.key)
```
Comment (blacha) [inline scripts/files/fs_s3.py:133]: why is the source object recreated here? 

is it not the same object as https://github.com/linz/topo-imagery/pull/195/files#diff-3874c4d626707da994208cc5bced4772a5d0506c9b9912a63ff645f0b05450e7R104

File: scripts/files/fs_s3.py [hunk 3]
```diff
@@ -67,6 +67,96 @@ def read(path: str, needs_credentials: bool = False) -> bytes:
...
+            path=path,
+            destination=new_path,
+            error="The source and destination path are not in the same bucket",
+        )
+        raise Exception(f"Files {path} and {new_path} are not on the same S3 bucket.")
+
+    s3 = boto3.resource("s3")
+    try:
+        if needs_credentials:
+            s3 = get_session(path).client("s3")
```
Comment (blacha) [inline scripts/files/fs_s3.py:101]: this is only going to work if the source and targets are the same bucket?

File: scripts/files/fs_s3.py (hunk 2 above)
Comment (paulfouquet) [reply scripts/files/fs_s3.py:133]: I forgot to remove that as I did push a commit later to check if the source object exists

File: scripts/files/fs_s3.py (hunk 3 above)
Comment (paulfouquet) [reply scripts/files/fs_s3.py:101]: Yes

File: scripts/files/fs_local.py (hunk 1 above)
Comment (paulfouquet) [reply scripts/files/fs_local.py:37]: Looks like it does manage those: https://docs.python.org/3/library/os.html#os.rename

Comment (paulfouquet) [reply scripts/files/fs_local.py:37]: > does it also throw if the target does exist?
I was looking at this now and

It does override the file as per
> If both are files, dst will be replaced silently if the user has permission.

I'm not too sure of what behaviour we want to have for the general functionality? 
I think for our context of renaming the tiff files as we don't want any duplicate we should probably do a check of duplicate files before the process of validate the tiff anyway?

**Pull request closed without merging**